newIDE/app/src/UI/Carousel.js

```diff
@@ -0,0 +1,356 @@
+  const [
+    shouldDisplayLeftArrow,
+    setShouldDisplayLeftArrow,
+  ] = React.useState<boolean>(false);
+  const classesForGridList = useStylesForGridList();
+  const classesForGridListItem = useStylesForGridListItem();
+  const scrollView = React.useRef<?HTMLUListElement>(null);
+  const areItemsSet = items && items.length;
+  const itemsToDisplay =
+    items && items.length
+      ? items
+      : Array(3)
+          .fill({
+            skeleton: true,
+            title: '',
+            description: '',
+            thumbnail: '',
+          })
+          .map((item, index) => ({ ...item, id: `skeleton${index}` }));
+
+  const windowSize = useResponsiveWindowWidth();
+  const imageHeight = referenceSizesByWindowSize.imageHeight[windowSize];
+  const arrowWidth = referenceSizesByWindowSize.arrowWidth[windowSize];
+  const cellWidth = (16 / 9) * imageHeight;
+  const widthUnit = cellWidth + cellSpacing;
+
+  const cellHeight =
+    imageHeight + (displayItemTitles ? titleHeight + spacerSize : 0);
+
+  const renderImage = React.useCallback(
+    (item: ThumbnailType | SkeletonThumbnail): React.Node => (
+      <img
+        src={item.thumbnailUrl}
+        style={{
+          ...styles.image,
+          height: imageHeight,
+          minHeight: imageHeight,
+          width: cellWidth,
+        }}
+        alt={item.title}
+        title={item.title}
+      />
+    ),
+    [cellWidth, imageHeight]
+  );
+
+  const renderThumbnail = React.useCallback(
+    (item: ThumbnailType | SkeletonThumbnail): ?React.Node => {
+      if (!item.skeleton && !item.link && !item.thumbnailUrl) return null;
+      if (item.link) {
+        return (
+          <a href={item.link} target="_blank">
+            {renderImage(item)}
+          </a>
+        );
+      }
+      if (item.thumbnailUrl) {
+        return renderImage(item);
+      }
+      if (item.skeleton) {
+        return (
+          <Skeleton variant="rect" height={imageHeight} width={cellWidth} />
+        );
+      }
+    },
+    [renderImage, cellWidth, imageHeight]
+  );
+
+  const renderItemTitle = React.useCallback(
+    (item: ThumbnailType | SkeletonThumbnail): ?React.Node => {
+      if (!displayItemTitles) return null;
+      return (
+        <>
+          <Spacer />
+          {item.title ? (
+            <div style={{ ...styles.itemTitleContainer, width: cellWidth }}>
+              <Text noMargin style={styles.itemTitle}>
+                {item.title}
+              </Text>
+            </div>
+          ) : (
+            <Skeleton
+              variant="rect"
+              height={titleHeight}
+              width={(cellWidth / 3) * (1 + 2 * Math.random())} // Make rectangles of different lengths so that the UI feels more "alive".
+            />
+          )}
+        </>
+      );
+    },
+    [cellWidth, titleHeight]
+  );
+
+  const renderBrowseAllButton = (): React.Node => {
+    const flatButtonProps: $Shape<FlatButtonProps> = { onClick: () => {} };
```

Comment:
To avoid using $Shape, I would create the onClick function first, with `let onClick = () => {}` then assign it, then use it in the FlatButton directly ?
`<FlatButton onClick={onClick} ... />`

Reply:
Or just a bunch of ifs inside the prop if really it's super complicated ;) 
`onClick={() => { if (...) else (...) }}`